File: cv2.jl

include("OpenCVImages.jl")

module cv2
    using CxxWrap
    @wrapmodule(joinpath("lib","libcv2_jl"), :cv2)

    function __init__()
        @initcxx
    end
    import Main.OpenCVImages


    const CV_CN_MAX = 512
    const CV_CN_SHIFT = 3
    const CV_DEPTH_MAX = (1 << CV_CN_SHIFT)

    const CV_8U = 0
    const CV_8S = 1
    const CV_16U = 2
    const CV_16S = 3
    const CV_32S = 4
    const CV_32F = 5
    const CV_64F = 6

    const CV_MAT_DEPTH_MASK = (CV_DEPTH_MAX - 1)
    CV_MAT_DEPTH(flags) = ((flags) & CV_MAT_DEPTH_MASK)

    CV_MAKETYPE(depth,cn) = (CV_MAT_DEPTH(depth) + (((cn)-1) << CV_CN_SHIFT))
    CV_MAKE_TYPE = CV_MAKETYPE
 

    struct KeyPoint
        pt::Tuple{Float32,Float32}
        size::Float32
        angle::Float32
        response::Float32
        octave::Integer
        class_id::Integer
    end

    const Image = Union{OpenCVImages.OpenCVImage{A} where {A}, SubArray{UInt8, N, OpenCVImages.OpenCVImage{A}, T} where {N, A, T}}
    const Scalar = Union{Tuple{Float64}, Tuple{Float64, Float64}, Tuple{Float64, Float64, Float64}, NTuple{4, Float64}}
    const Size = Tuple{Integer, Integer}
    const Rect = NTuple{4, Integer}

    const size_t = UInt64 #TODO: Get size_t from CxxWrap maybe

    const CASCADE_SCALE_IMAGE = Integer(2)
    const COLOR_BGR2GRAY = Integer(6)

    change_major_order(X::AbstractArray, size...=size(X)...) = permutedims(reshape(X, reverse([size...])...), length(size):-1:1)

    function cpp_mat_to_jl_arr(mat)
        arr = cv2.Mat_mutable_data(mat)
        #TODO: Implement types
        #Preserve Mat so that array allocated by C++ isn't deallocated
        return OpenCVImages.OpenCVImage{UInt8}(mat, arr)
    end

    function jl_arr_to_cpp_mat(img::Image)
        # TODO: UserTypes do not work with StridedArray. Find something else.
        steps = strides(img)
        if steps[1] <= steps[2] <= steps[3] && steps[1]==1
            steps_a = Array{size_t, 1}()
            ndims_a = Array{Int32, 1}()
            sz = sizeof(eltype(img))
            push!(steps_a, UInt64(steps[3]*sz))
            push!(steps_a, UInt64(steps[2]*sz))
            push!(steps_a, UInt64(steps[1]*sz))

            push!(ndims_a, Int32(size(img)[3]))
            push!(ndims_a, Int32(size(img)[2]))

            return cv2.Mat(2, pointer(ndims_a), CV_MAKE_TYPE(CV_8U, size(img)[1]), Ptr{Nothing}(pointer(img)), pointer(steps_a))
        else
            print("Bad steps")
            print(steps)
        end
    end

    function jl_pt_to_cpp_pt(pt::Tuple{Real, Real})
        return cv2.Point2f(Float32(pt[1]), Float32(pt[2]))
    end

    function jl_scalar_to_cpp_scalar(sc::Scalar)
        if size(sc,1)==1
            return cv2.Scalar_(sc[1], 0, 0, 0)
        elseif size(sc,1) == 2
            return cv2.Scalar_(sc[1], sc[2], 0, 0)
        elseif size(sc,1) == 3
            return cv2.Scalar_(sc[1], sc[2], sc[3], 0)
        end
        return cv2.Scalar_(sc[1], sc[2], sc[3], sc[4])
    end

    function jl_size_to_cpp_size(sx::Size)
        return cv2.Size2i(sx[1], sx[2])
    end

    function cpp_Point2f_to_jl_tuple(pt)
        return (cv2.Point2f_get_x(pt), cv2.Point2f_get_y(pt))
    end

    function cpp_KeyPoint_to_jl_KeyPoint(kp)
        kpr = KeyPoint(cpp_Point2f_to_jl_tuple(cv2.KeyPoint_get_pt(kp)), cv2.KeyPoint_get_size(kp), 
                        cv2.KeyPoint_get_angle(kp), cv2.KeyPoint_get_response(kp), 
                        cv2.KeyPoint_get_octave(kp), cv2.KeyPoint_get_class_id(kp))
        return kpr
    end

    function cpp_Size2i_to_jl_tuple(sz)
        return (cv2.Size2i_get_width(sz), cv2.Size2i_get_height(sz))
    end

    function cpp_Rect2i_to_jl_tuple(pt)
        return (cv2.Rect2i_get_x(pt), cv2.Rect2i_get_y(pt), cv2.Rect2i_get_width(pt), cv2.Rect2i_get_height(pt))
    end
    function imshow(winname::String, img::Image)
        Mat_img = cv2.jl_arr_to_cpp_mat(img)
        cv2.imshow(winname, Mat_img)
    end
    function imread(filename::String, flags::Integer)
        Mat_tmp = cv2.imread_(filename, flags)
        jl_Image = cpp_mat_to_jl_arr(Mat_tmp)
        return jl_Image
    end
    function simpleBlobDetector_detect(algo, img::Image)
        ret = cv2.simpleBlobDetector_detect(algo, jl_arr_to_cpp_mat(img))
        arr = Array{cv2.KeyPoint, 1}()
        for it in ret
            push!(arr, cpp_KeyPoint_to_jl_KeyPoint(it))
        end
        return arr
    end

    function rectangle(img::Image, pt1::Tuple{Integer, Integer}, pt2::Tuple{Integer, Integer}, color::Scalar; thickness::Integer=1, lineType::Integer=8, shift::Integer=0)
        cv2.rectangle(jl_arr_to_cpp_mat(img), jl_pt_to_cpp_pt(pt1), jl_pt_to_cpp_pt(pt2), jl_scalar_to_cpp_scalar(color), thickness, lineType, shift)
    end

    function VideoCapture_read(arg1)
        ret1 = cv2.read(arg1)
        return (ret1[1], cpp_mat_to_jl_arr(ret1[2]))
    end

    function cvtColor(src::Image, code::Integer)
        return cpp_mat_to_jl_arr(cv2.cvtColor(jl_arr_to_cpp_mat(src), code))
    end
    function equalizeHist(src::Image)
        return cpp_mat_to_jl_arr(cv2.equalizeHist(jl_arr_to_cpp_mat(src)))
    end

    function CascadeClassifier_detectMultiScale(inp1, image::Image;
                                             scaleFactor::Float64 = 1.1,
                                             minNeighbors::Integer = 3, 
                                             flags::Integer = 0,
                                             minSize::Size = (0,0),
                                             maxSize::Size = (0,0))
        ret = cv2.detectMultiScale(inp1, jl_arr_to_cpp_mat(image), scaleFactor, minNeighbors, flags, jl_size_to_cpp_size(minSize), jl_size_to_cpp_size(maxSize))
        arr = Array{Rect, 1}()
        for it in ret
            push!(arr, cpp_Rect2i_to_jl_tuple(it))
        end
        return arr
    end

end

File: cv2.cpp
#include <vector>

#include "jlcxx/jlcxx.hpp"
#include "jlcxx/functions.hpp"
#include "jlcxx/stl.hpp"
#include "jlcxx/array.hpp"
#include "jlcxx/tuple.hpp"

#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/features2d.hpp>
#include <opencv2/core/utility.hpp>
#include <opencv2/objdetect.hpp>
#include <opencv2/imgproc.hpp>
#include "opencv2/highgui.hpp"
#include "opencv2/videoio.hpp"

using namespace cv;
using namespace std;
using namespace jlcxx;

namespace jlcxx
{
template <>
struct SuperType<SimpleBlobDetector>
{
    typedef Feature2D type;
};

template <typename T>
struct IsSmartPointerType<cv::Ptr<T>> : std::true_type
{
};
template <typename T>
struct ConstructorPointerType<cv::Ptr<T>>
{
    typedef T *type;
};
} // namespace jlcxx

jlcxx::ArrayRef<uint8_t, 3> Mat_mutable_data(cv::Mat Mat)
{
    return make_julia_array(Mat.data, Mat.channels(), Mat.cols, Mat.rows);
}

JLCXX_MODULE define_julia_module(jlcxx::Module &mod)
{
    mod.add_type<Mat>("Mat").constructor<int, const int*, int, void *, const size_t*>();

    mod.add_type<Point2f>("Point2f").constructor<float, float>().method("Point2f_get_x", [](const Point2f &a) { return a.x; }).method("Point2f_get_y", [](const Point2f &a) { return a.y; });

    mod.add_type<Size2i>("Size2i").constructor<int, int>().method("Size2i_get_height", [](const Size2i &a) {return a.height;}).method("Size2i_get_width",[](const Size2i &a){return a.width;});
    mod.add_type<Rect2i>("Rect2i").constructor<int, int, int, int>()
                                        .method("Rect2i_get_height", [](const Rect2i &a) {return a.height;})
                                        .method("Rect2i_get_width",[](const Rect2i &a){return a.width;})
                                        .method("Rect2i_get_x",[](const Rect2i &a){return a.x;})
                                        .method("Rect2i_get_y",[](const Rect2i &a){return a.y;});

    mod.add_type<Scalar>("Scalar").constructor<double, double, double, double>();
    mod.add_type<KeyPoint>("KeyPoint").method("KeyPoint_get_pt", [](const KeyPoint &a) { return a.pt; }).method("KeyPoint_get_size", [](const KeyPoint &a) { return a.size; }).method("KeyPoint_get_angle", [](const KeyPoint &a) { return a.angle; }).method("KeyPoint_get_response", [](const KeyPoint &a) { return a.response; }).method("KeyPoint_get_octave", [](const KeyPoint &a) { return a.octave; }).method("KeyPoint_get_class_id", [](const KeyPoint &a) { return a.class_id; });
    mod.add_type<Feature2D>("Feature2D");
    mod.add_type<SimpleBlobDetector>("SimpleBlobDetector", jlcxx::julia_base_type<Feature2D>());

    mod.add_type<CascadeClassifier>("CascadeClassifier").constructor<const std::string &>().method("detectMultiScale", [](CascadeClassifier c1, Mat a1, double a2, int a3, int a4, Size2i a5, Size2i a6){    std::vector<Rect> o1;c1.detectMultiScale(a1, o1,a2,a3,a4,a5,a6);return o1;}).method("empty", &CascadeClassifier::empty);
    mod.add_type<VideoCapture>("VideoCapture").constructor<const std::string &>().constructor<int>().method("read", [](VideoCapture c1) { Mat dst; return make_tuple(c1.read(dst), dst); });

    jlcxx::add_smart_pointer<cv::Ptr>(mod, "cv_Ptr");

    mod.method("imread", [](string a1, int a2) { return cv::imread(a1, a2); });
    mod.method("imshow", [](string a1, Mat a2) { return cv::imshow(a1, a2); });
    mod.method("namedWindow", [](string a1) { cv::namedWindow(a1); });
    mod.method("namedWindow", [](string a1, int a2) { cv::namedWindow(a1, a2); });
    mod.method("waitKey", [](int a1) { return cv::waitKey(a1); });
    mod.method("rectangle", [](Mat a1, Point2f a2, Point2f a3, const Scalar &a4, int a5, int a6, int a7) { cv::rectangle(a1, a2, a3, a4, a5, a6, a7); });
    mod.method("cvtColor", [](Mat a1, int a2) {Mat o1; cv::cvtColor(a1, o1, a2); return o1; });
    mod.method("equalizeHist", [](Mat a1) {Mat o1; cv::equalizeHist(a1, o1); return o1; });
    mod.method("destroyAllWindows", []() { cv::destroyAllWindows(); });


    mod.method("Mat_mutable_data", &Mat_mutable_data);
    mod.method("Mat_mutable_data_2", [](Mat m){
        return make_tuple(m.data, m.type(), m.channels(), m.size[1], m.size[0], m.step[1], m.step[0]);
    });

    // Algorithm Inherits
    mod.method("simpleBlobDetector_detect", [](cv::Ptr<SimpleBlobDetector> c1, Mat a1) { vector<KeyPoint> o1; c1->detect(a1, o1); return o1; });
    mod.method("simpleBlobDetector_create", []() { return cv::Ptr<SimpleBlobDetector>(SimpleBlobDetector::create()); });


    //callback work
    mod.method("createButton", [](const string & bar_name, jl_function_t* on_change, int type, bool initial_button_state) {createButton(bar_name, [](int s, void* c) {
        JuliaFunction f((jl_function_t*)c);
        f(forward<int>(s));
    }, (void*)on_change, type, initial_button_state);});

    mod.method("setMouseCallback", [](const string & winname, jl_function_t* onMouse) {
        setMouseCallback(winname, [](int event, int x, int y, int flags, void* c) {
        JuliaFunction f((jl_function_t*)c);
        f(forward<int>(event), forward<int>(x), forward<int>(y), forward<int>(flags));
    }, (void*)onMouse);});

    mod.method("createTrackbar", [](const String &trackbarname, const String &winname, int& value, int count, jl_function_t* onChange) {
        createTrackbar(trackbarname, winname, &value, count, [](int s, void* c) {
        JuliaFunction f((jl_function_t*)c);
        f(forward<int>(s));
    }, (void*)onChange);});


}
